x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/CheckTargetShardsCountStep.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+package org.elasticsearch.xpack.core.ilm;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.common.xcontent.ParseField;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.index.Index;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * This step checks whether the new shrunken index's shards count is a factor of the source index's shards count.
+ */
+public class CheckTargetShardsCountStep extends ClusterStateWaitStep {
+
+    public static final String NAME = "check-target-shards-count";
+
+    private final Integer numberOfShards;
+
+    private static final Logger logger = LogManager.getLogger(CheckTargetShardsCountStep.class);
+
+    CheckTargetShardsCountStep(StepKey key, StepKey nextStepKey, Integer numberOfShards) {
+        super(key, nextStepKey);
+        this.numberOfShards = numberOfShards;
+    }
+
+    @Override
+    public boolean isRetryable() {
+        return true;
+    }
+
+    public Integer getNumberOfShards() {
+        return numberOfShards;
+    }
+
+    @Override
+    public Result isConditionMet(Index index, ClusterState clusterState) {
+        IndexMetadata indexMetadata = clusterState.metadata().index(index);
```

Comment:
Please check for null `indexMetadata` (in case the index was deleted in the meantime) 

eg.
```
        if (indexMetadata == null) {
            // Index must have been since deleted, ignore it
            logger.debug("[{}] lifecycle action for index [{}] executed but index no longer exists",
                getKey().getAction(), index.getName());
            return new Result(false, null);
        }
```